model/riscv_insts_zcmp.sail

```diff
@@ -0,0 +1,487 @@
+        0b1010 => 32,
+        0b1011 => 32,
+        0b1100 => 48,
+        0b1101 => 48,
+        0b1110 => 48,
+        0b1111 => 64,
+        _ => internal_error(__FILE__, __LINE__, "Unsupported rlist: " ^ bits_str(rlist))
+    }
+  else {
+    match(rlist) {
+        0b0100 => 16,
+        0b0101 => 16,
+        0b0110 => 32,
+        0b0111 => 32,
+        0b1000 => 48,
+        0b1001 => 48,
+        0b1010 => 64,
+        0b1011 => 64,
+        0b1100 => 80,
+        0b1101 => 80,
+        0b1110 => 96,
+        0b1111 => 112,
+        _ => internal_error(__FILE__, __LINE__, "Unsupported rlist: " ^ bits_str(rlist))
+    }
+  }
+}
+
+mapping reg_list : bits(4) <-> string = {
+  0b0100     <-> "ra",
+  0b0101     <-> "ra, s0",
+  0b0110     <-> "ra, s0-s1",
+  0b0111     <-> "ra, s0-s2",
+  0b1000     <-> "ra, s0-s3",
+  0b1001     <-> "ra, s0-s4",
+  0b1010     <-> "ra, s0-s5",
+  0b1011     <-> "ra, s0-s6",
+  0b1100     <-> "ra, s0-s7",
+  0b1101     <-> "ra, s0-s8",
+  0b1110     <-> "ra, s0-s9",
+  0b1111     <-> "ra, s0-s11",
+}
+
+mapping rlist_to_int : bits(4) <-> int = {
+  0b0000     <-> 0,
+  0b0001     <-> 1,
+  0b0010     <-> 2,
+  0b0011     <-> 3,
+  0b0100     <-> 4,
+  0b0101     <-> 5,
+  0b0110     <-> 6,
+  0b0111     <-> 7,
+  0b1000     <-> 8,
+  0b1001     <-> 9,
+  0b1010     <-> 10,
+  0b1011     <-> 11,
+  0b1100     <-> 12,
+  0b1101     <-> 13,
+  0b1110     <-> 14,
+  0b1111     <-> 15,
+}
+
+mapping spimm_to_int : bits(2) <-> int = {
+  0b00     <-> 0,
+  0b01     <-> 1,
+  0b10     <-> 2,
+  0b11     <-> 3,
+}
+
+function zcmp_regmask(rlist : bits(4)) -> vector(32, bool) = {
+  var mask : vector(32, bool) = undefined;
+
+  if(unsigned(rlist) >= unsigned(0b100)) then {
+    mask[1] = true
+  };
+
+  let rlist_num = rlist_to_int(rlist);
+
+  foreach (i from 5 to rlist_num) {
+    assert(i <= 15);
+    if( (i - 5) < 2 ) then
+      mask[3 + i] = true
+    else {
+      mask[11 + i] = true
+    };
+  };
+
+  if(rlist == zero_extend(0xf)) then
+    mask[27] = true;
+
+  mask
+}
+
+enum clause extension = Ext_Zcmp
+function clause extensionEnabled(Ext_Zcmp) = extensionEnabled(Ext_Zca) & extensionEnabled(Ext_Zcd) == false & (xlen == 32 | xlen == 64)
+
+union clause ast = CM_PUSH : (bits(4), bits(2))
+
+mapping clause encdec_compressed = CM_PUSH(rlist, spimm)   if extensionEnabled(Ext_Zcmp)
+  <-> 0b101 @ 0b11000 @ rlist : bits(4) @ spimm : bits(2) @ 0b10    if extensionEnabled(Ext_Zcmp)
+
+val process_cmpush : (bits(4), bits(2), word_width) -> Retired
+function process_cmpush (rlist : bits(4), spimm : bits(2), width : word_width) = {
+  let width_bytes = size_bytes(width);
+  assert(width_bytes <= xlen_bytes);
+
+  let addr = X(sp);
+  let spimm_int = spimm_to_int(spimm);
+  let stack_adj = negate_int(get_stack_adj_base(rlist) + spimm_int * 16);
+  let new_sp = addr + to_bits(xlen, stack_adj);
+  let mask = zcmp_regmask(rlist);
+  let bytes = if sizeof(xlen) == 32 then 4 else 8;
+
+  var imm : int = width_bytes;
```

Comment:
You can do `var imm : xlenbits = zeros();` and `imm = imm - width_bytes`.

I think `imm` is the wrong name too surely? Did you mean `offset`?